### PR TITLE
yarn: Universal utility for Yarn version detection

### PR DESCRIPTION
**Pull request by a-ovchinnikov**
This commit unifies handling of version info for Yarn and Yarn Classic package managers.

Resolves: https://github.com/containerbuildsystem/cachi2/issues/660

# Maintainers will complete the following section

- [ ] Commit messages are descriptive enough
- [ ] Code coverage from testing does not decrease and new code is covered
- [ ] Docs updated (if applicable)
- [ ] Docs links in the code are still valid (if docs were updated)

**Note:** if the contribution is external (not from an organization member), the CI
pipeline will not run automatically. After verifying that the CI is safe to run:

- [approve GitHub Actions workflows][approve-gh-actions] by clicking a button
- approve the Red Hat Trusted App Pipeline container build by commenting `/ok-to-test`
  (as is the [standard for Pipelines as Code][pac-running-pipeline])

[approve-gh-actions]: https://docs.github.com/en/actions/managing-workflow-runs/approving-workflow-runs-from-public-forks
[pac-running-pipeline]: https://pipelinesascode.com/docs/guide/running/#running-the-pipelinerun

### CODE REVIEW

File: cachi2/core/package_managers/yarn/utils.py
Comment (slimreaper35) [inline cachi2/core/package_managers/yarn/utils.py:1]: Is this the correct location for this class/function ?

File: cachi2/core/package_managers/yarn/utils.py [hunk 1]
```diff
@@ -27,3 +29,68 @@ def run_yarn_cmd(
     except subprocess.CalledProcessError as e:
         # the yarn command writes the errors to stdout
         raise PackageManagerError(f"Yarn command failed: {' '.join(cmd)}", stderr=e.stdout)
+
+
+VersionLike = Union[Version, str]
+
+
+class Versions_range:
```
Comment (slimreaper35) [inline cachi2/core/package_managers/yarn/utils.py:37]:
```suggestion
class VersionRange:
```

Comment (a-ovchinnikov) [reply cachi2/core/package_managers/yarn/utils.py:37]: Yes! Thank you!

File: cachi2/core/package_managers/yarn/utils.py
Comment (a-ovchinnikov) [reply cachi2/core/package_managers/yarn/utils.py:1]: For the time being yes, but I am open to suggestions.

Comment (slimreaper35) [reply cachi2/core/package_managers/yarn/utils.py:1]: Maybe it depends on https://github.com/containerbuildsystem/cachi2/issues/624

File: cachi2/core/package_managers/yarn/utils.py [hunk 2]
```diff
@@ -27,3 +29,68 @@ def run_yarn_cmd(
...
+        """Initialize a version range."""
+        self.min_ver = min_ver if isinstance(min_ver, Version) else Version.parse(min_ver)
+        self.max_ver = max_ver if isinstance(max_ver, Version) else Version.parse(max_ver)
+
+    def __contains__(self, other: Version) -> bool:
+        if not isinstance(other, self.min_ver.__class__):
+            return False
+        # The original version check logic (as captured in UTs) broke with direct
+        # version comparison rules, e.g. 4.0.0.rc1 would have been considered
+        # version 4 and would  have been rejected basing on major version
```
Comment (eskultety) [inline cachi2/core/package_managers/yarn/utils.py:70]: This was intentional and I'd argue that it's even correct since RCs are like ~90% of the final release and so if we're not ready for the next major we can't be ready even for the the RC in terms of support, so how exactly did it break UTs?

File: cachi2/core/package_managers/yarn/utils.py [hunk 3]
```diff
@@ -27,3 +29,68 @@ def run_yarn_cmd(
     except subprocess.CalledProcessError as e:
         # the yarn command writes the errors to stdout
         raise PackageManagerError(f"Yarn command failed: {' '.join(cmd)}", stderr=e.stdout)
+
+
+VersionLike = Union[Version, str]
```
Comment (eskultety) [inline cachi2/core/package_managers/yarn/utils.py:34]: nitpick: `packaging.Version` is different from `semver.Version`. We do use **a** `Version` object interchangeably throughout the code base IIRC so a type name of `SemVerLike` feels more appropriate for the casual code reader.

Even I missed the `semver` import for the first couple of minutes of this review and was puzzled on how would this close #660, until I found the import.

File: cachi2/core/package_managers/yarn/utils.py [hunk 4]
```diff
@@ -27,3 +29,68 @@ def run_yarn_cmd(
...
+            return False
+        # The original version check logic (as captured in UTs) broke with direct
+        # version comparison rules, e.g. 4.0.0.rc1 would have been considered
+        # version 4 and would  have been rejected basing on major version
+        # only. The original implementation considered anything starting with
+        # a different major version an outlier. The logic below captures this
+        # with a special case for versions with prerelease field set.
+        if other.prerelease:
+            # Drop prerelease:
+            other_prime = Version(other.major, other.minor, other.patch)
```
Comment (eskultety) [inline cachi2/core/package_managers/yarn/utils.py:76]: nitpick: what is the meaning of "prime" here? I'm a bit lost in translation; wouldn't `other_ver` be less linguistically challenging :)?

File: cachi2/core/package_managers/yarn/main.py [hunk 5]
```diff
@@ -228,19 +232,14 @@ def _generate_environment_variables() -> list[EnvironmentVariable]:
...
-    installed_yarn_version = run_yarn_cmd(
-        ["--version"], source_dir, env={"COREPACK_ENABLE_DOWNLOAD_PROMPT": "0"}
-    ).strip()
-    try:
-        if installed_yarn_version != expected_version:
-            raise PackageManagerError(
-                f"Cachi2 expected corepack to install yarn@{expected_version} but instead "
-                f"found yarn@{installed_yarn_version}."
-            )
-    except ValueError as exc:
```
Comment (eskultety) [inline cachi2/core/package_managers/yarn/main.py:240]: This code extraction should be a separate commit.

**Review by eskultety — COMMENTED** [review]
Nitpicks only, quick respin will do!

File: cachi2/core/package_managers/yarn/utils.py (hunk 4 above)
Comment (a-ovchinnikov) [reply cachi2/core/package_managers/yarn/utils.py:76]: I would have written it as `other'` if Python allowed me to, I have spelled `'` out without thinking twice. I'll update to `other_ver` to make this look less surprising. _Edit_: `öther` is a tempting variant too, but I'll restrain myself.

File: cachi2/core/package_managers/yarn/utils.py (hunk 3 above)
Comment (a-ovchinnikov) [reply cachi2/core/package_managers/yarn/utils.py:34]: Done.

File: cachi2/core/package_managers/yarn/main.py (hunk 5 above)
Comment (a-ovchinnikov) [reply cachi2/core/package_managers/yarn/main.py:240]: Done.

File: cachi2/core/package_managers/yarn/utils.py [hunk 6]
```diff
@@ -1,6 +1,8 @@
 import os
 import subprocess
-from typing import Optional
+from typing import Optional, Union
```
Comment (eskultety) [inline cachi2/core/package_managers/yarn/utils.py:3]: **Commit message**:

s/unifies/unified
s/handling of/handling of Yarn
s/and/with a new

File: cachi2/core/package_managers/yarn/main.py (hunk 5 above)
Comment (eskultety) [reply cachi2/core/package_managers/yarn/main.py:240]: I'm not sure about the extraction as proposed, i.e.:

1. commit 1 adds a duplicate handler (with a single line changed)
2. commit 2 removes the original code no longer used

instead of
1. move the code to a new standalone helper
2. rest of the changes including class definition

otherwise it's not proposed as an extraction IMO (as a logically isolated change).

File: cachi2/core/package_managers/yarn/main.py [hunk 7]
```diff
@@ -13,7 +13,11 @@
     get_semver_from_yarn_path,
 )
 from cachi2.core.package_managers.yarn.resolver import create_components, resolve_packages
-from cachi2.core.package_managers.yarn.utils import run_yarn_cmd
+from cachi2.core.package_managers.yarn.utils import (
```
Comment (eskultety) [inline cachi2/core/package_managers/yarn/main.py:16]: This has an identical commit subject as the previous commit, was that intentional? It may look weird in the condensed git history dump so one of them should be tweaked ever so slightly.

File: cachi2/core/package_managers/yarn/utils.py (hunk 6 above)
Comment (a-ovchinnikov) [reply cachi2/core/package_managers/yarn/utils.py:3]: Done.

File: cachi2/core/package_managers/yarn/main.py (hunk 7 above)
Comment (a-ovchinnikov) [reply cachi2/core/package_managers/yarn/main.py:16]: No, it was not. Fixed.

File: cachi2/core/package_managers/yarn/main.py (hunk 5 above)
Comment (a-ovchinnikov) [reply cachi2/core/package_managers/yarn/main.py:240]: Good point, I'll restructure it.

**Review by slimreaper35 — APPROVED** [review]

File: cachi2/core/package_managers/yarn/main.py (hunk 7 above)
Comment (slimreaper35) [reply cachi2/core/package_managers/yarn/main.py:16]: is it ? or I am blind

**Review by taylormadore — APPROVED** [review]

File: cachi2/core/package_managers/yarn/main.py (hunk 7 above)
Comment (a-ovchinnikov) [reply cachi2/core/package_managers/yarn/main.py:16]: Not yet, I am still re-shaping the two commits, but it will be fixed in the next push

Comment (a-ovchinnikov) [reply cachi2/core/package_managers/yarn/main.py:16]: Now should be done.

File: cachi2/core/package_managers/yarn/main.py (hunk 5 above)
Comment (eskultety) [reply cachi2/core/package_managers/yarn/main.py:240]: I may not have been very clear on this especially using the whole hunk as context. Admittedly, there was a slight oversight on my side regarding the version mismatch check which stays in this function, so per our private conversation with @a-ovchinnikov I'll just record the gist of what I meant by splicing that bit of logic:
```
# main.py
-    installed_yarn_version = run_yarn_cmd(
-        ["--version"], source_dir, env={"COREPACK_ENABLE_DOWNLOAD_PROMPT": "0"}
+    installed_yarn_version = extract_yarn_version_from_env(source_dir)
...

# utils.py
+def extract_yarn_version_from_env(source_dir: RootedPath) -> Version:
+    """Extract yarn version from environment."""
+    yarn_version_output = run_yarn_cmd(["--version"], source_dir).strip()
+    return yarn_version_output
```
and then the rest of the changes essentially.

File: cachi2/core/package_managers/yarn/main.py [hunk 8]
```diff
@@ -228,19 +232,14 @@ def _generate_environment_variables() -> list[EnvironmentVariable]:
...
-    try:
-        if installed_yarn_version != expected_version:
-            raise PackageManagerError(
-                f"Cachi2 expected corepack to install yarn@{expected_version} but instead "
-                f"found yarn@{installed_yarn_version}."
-            )
-    except ValueError as exc:
+    installed_yarn_version = extract_yarn_version_from_env(
+        source_dir,
+        env={"COREPACK_ENABLE_DOWNLOAD_PROMPT": "0"},
```
Comment (eskultety) [inline cachi2/core/package_managers/yarn/main.py:237]: We used an explicit env because we were calling into the Yarn cmd executor wrapper directly. That is now abstracted by the new version parser helper and so the helper can hardcode this bit, no need to pass the env to a single-purpose helper as an argument :).

Comment (a-ovchinnikov) [reply cachi2/core/package_managers/yarn/main.py]: Done.

File: cachi2/core/package_managers/yarn/main.py (hunk 5 above)
Comment (a-ovchinnikov) [reply cachi2/core/package_managers/yarn/main.py:240]: Oh, now I see it, thank you! Done.

File: cachi2/core/package_managers/yarn/main.py [hunk 9]
```diff
@@ -118,7 +118,7 @@ def _configure_yarn_version(project: Project) -> None:
     yarn_path_version = get_semver_from_yarn_path(project.yarn_rc.yarn_path)
     package_manager_version = get_semver_from_package_manager(project.package_json.package_manager)
 
-    if not yarn_path_version and not package_manager_version:
```
Comment (eskultety) [inline cachi2/core/package_managers/yarn/main.py:121]: **Commit msg**:
s/int/into

File: cachi2/core/package_managers/yarn/main.py [hunk 10]
```diff
@@ -118,7 +118,7 @@ def _configure_yarn_version(project: Project) -> None:
...
+    if yarn_path_version is None and package_manager_version is None:
```
Comment (eskultety) [inline cachi2/core/package_managers/yarn/main.py:125]: Unrelated PEP correctness check update --> separate commit.

**Review by eskultety — APPROVED** [review]
ACK with the tiny adjustments.

File: cachi2/core/package_managers/yarn/main.py (hunk 9 above)
Comment (a-ovchinnikov) [reply cachi2/core/package_managers/yarn/main.py:121]: Done.

File: cachi2/core/package_managers/yarn/main.py (hunk 10 above)
Comment (a-ovchinnikov) [reply cachi2/core/package_managers/yarn/main.py:125]: Ok. Done.
